src/jvm/main/org/jetbrains/kotlinx/lincheck/runner/TestThreadExecutionGenerator.java

```diff
@@ -256,7 +265,6 @@ private static void generateRun(ClassVisitor cv, Type testType, int iThread, Lis
             mv.visitLabel(unexpectedExceptionHandler);
             // Call onFailure method
             mv.dup();
```

Comment:
You can add a special "failOnExceptionIsUnexpected(e: Throwable)" method that
1) checks whether the exception cannot be a result
2) if everything is fine, finishes 
3) if not, calls Runner.onFailure(..) method and throws the exception.

This way, you need neither this bytecode generation nor the added in this PR one.